.github/ci-heapster.sh

```diff
@@ -0,0 +1,152 @@
+  name="$(basename "$exe")"
+  echo "Copying $name to $2"
+  mkdir -p "$2"
+  cp -f "$exe" "$2/$name"
+  $IS_WIN || chmod +x "$2/$name"
+}
+
+retry() {
+  echo "Attempting with retry:" "$@"
+  local n=1
+  while true; do
+    if "$@"; then
+      break
+    else
+      if [[ $n -lt 3 ]]; then
+        sleep $n # don't retry immediately
+        ((n++))
+        echo "Command failed. Attempt $n/3:"
+      else
+        echo "The command has failed after $n attempts."
+        return 1
+      fi
+    fi
+  done
+}
+
+setup_dist_bins() {
+  is_exe "dist/bin" "saw" && is_exe "dist/bin" "jss" && return
+  extract_exe "saw" "dist/bin"
+  extract_exe "jss" "dist/bin"
+  export PATH=$PWD/dist/bin:$PATH
+  echo "$PWD/dist/bin" >> $GITHUB_PATH
+  strip dist/bin/saw* || echo "Strip failed: Ignoring harmless error"
+  strip dist/bin/jss* || echo "Strip failed: Ignoring harmless error"
+}
+
+install_z3() {
+  is_exe "$BIN" "z3" && return
+
+  case "$RUNNER_OS" in
+    Linux) file="ubuntu-16.04.zip" ;;
+    macOS) file="osx-10.14.6.zip" ;;
+    Windows) file="win.zip" ;;
+  esac
+  curl -o z3.zip -sL "https://github.com/Z3Prover/z3/releases/download/z3-$Z3_VERSION/z3-$Z3_VERSION-x64-$file"
+
+  if $IS_WIN; then 7z x -bd z3.zip; else unzip z3.zip; fi
+  cp z3-*/bin/z3$EXT $BIN/z3$EXT
+  $IS_WIN || chmod +x $BIN/z3
+  rm z3.zip
+}
+
+install_cvc4() {
+  is_exe "$BIN" "cvc4" && return
+  version="${CVC4_VERSION#4.}" # 4.y.z -> y.z
+
+  case "$RUNNER_OS" in
+    Linux) file="x86_64-linux-opt" ;;
+    Windows) file="win64-opt.exe" ;;
+    macOS) file="macos-opt" ;;
+  esac
+  # Temporary workaround
+  if [[ "$RUNNER_OS" == "Linux" ]]; then
+    curl -o cvc4$EXT -sL "https://cvc4.cs.stanford.edu/downloads/builds/x86_64-linux-opt/unstable/cvc4-2020-08-18-x86_64-linux-opt"
+  else
+    curl -o cvc4$EXT -sL "https://github.com/CVC4/CVC4/releases/download/$version/cvc4-$version-$file"
+  fi
+  $IS_WIN || chmod +x cvc4$EXT
+  mv cvc4$EXT "$BIN/cvc4$EXT"
+}
+
+install_yices() {
+  is_exe "$BIN" "yices" && return
+  ext=".tar.gz"
+  case "$RUNNER_OS" in
+    Linux) file="pc-linux-gnu-static-gmp.tar.gz" ;;
+    macOS) file="apple-darwin18.7.0-static-gmp.tar.gz" ;;
+    Windows) file="pc-mingw32-static-gmp.zip" && ext=".zip" ;;
+  esac
+  curl -o "yices$ext" -sL "https://yices.csl.sri.com/releases/$YICES_VERSION/yices-$YICES_VERSION-x86_64-$file"
+
+  if $IS_WIN; then
+    7z x -bd "yices$ext"
+    mv "yices-$YICES_VERSION"/bin/*.exe "$BIN"
+  else
+    tar -xzf "yices$ext"
+    pushd "yices-$YICES_VERSION" || exit
+    sudo ./install-yices
+    popd || exit
+  fi
+  rm -rf "yices$ext" "yices-$YICES_VERSION"
+}
+
+install_yasm() {
+  is_exe "$BIN" "yasm" && return
+  if [[ "$RUNNER_OS" = "Linux" ]]; then
+    sudo apt-get update -q && sudo apt-get install -y yasm
+  else
+    brew install yasm
+  fi
+}
+
+build() {
+  cabal v2-update
+  cabal v2-configure -j2 --minimize-conflict-set
+  if ! retry cabal v2-build "$@" saw jss && [[ "$RUNNER_OS" == "macOS" ]]; then
```

Comment:
Because the github runners tend to be a bit flaky every now and then. The retry function is something I lifted verbatim from the rust language's CI setup a while ago and it's helped reduce spurious failures due to hiccups.

Reply:
OK.  From the CI perspective, that makes me cringe since builds should be reproducible and stable, but if it's needed for github runners I guess I'll just bite my tongue.